be/src/exprs/vectorized/array_functions.tpp

```diff
@@ -0,0 +1,104 @@
+        ColumnPtr dest_column = src_column->clone_empty();
+
+        HashSet hash_set;
+
+        if (columns[0]->is_nullable()) {
+            const auto* src_nullable_column = down_cast<const NullableColumn*>(src_column.get());
+            const auto* src_data_column = down_cast<const ArrayColumn*>(src_nullable_column->data_column().get());
+            auto& dest_nullable_column = down_cast<NullableColumn&>(*dest_column);
+            auto& dest_null_data = down_cast<NullableColumn&>(*dest_column).null_column_data();
+            auto& dest_data_column = down_cast<ArrayColumn&>(*dest_nullable_column.data_column());
+
+            dest_null_data = src_nullable_column->immutable_null_column_data();
+            dest_nullable_column.set_has_null(src_nullable_column->has_null());
+
+            if (src_nullable_column->has_null()) {
+                for (size_t i = 0; i < chunk_size; i++) {
+                    if (!src_nullable_column->is_null(i)) {
+                        _array_distinct_item<HashSet>(*src_data_column, i, &hash_set, &dest_data_column);
+                        hash_set.clear();
+                    } else {
+                        dest_data_column.append_default();
+                    }
+                }
+            } else {
+                for (size_t i = 0; i < chunk_size; i++) {
+                    _array_distinct_item<HashSet>(*src_data_column, i, &hash_set, &dest_data_column);
+                    hash_set.clear();
+                }
+            }
+        } else {
+            const auto* src_data_column = down_cast<const ArrayColumn*>(src_column.get());
+            auto* dest_data_column = down_cast<ArrayColumn*>(dest_column.get());
+
+            for (size_t i = 0; i < chunk_size; i++) {
+                _array_distinct_item<HashSet>(*src_data_column, i, &hash_set, dest_data_column);
+                hash_set.clear();
+            }
+        }
+        return dest_column;
+    }
+
+    template <typename HashSet>
+    static void _array_distinct_item(const ArrayColumn& column, size_t index, HashSet* hash_set,
+                                     ArrayColumn* dest_column) {
+        bool has_null = false;
+        // TODO: may be has performance problem, optimize later
+        Datum v = column.get(index);
+        const auto& items = v.get<DatumArray>();
+
+        for (const auto& item : items) {
+            if (item.is_null()) {
+                has_null = true;
+            } else {
+                hash_set->emplace(item.get<CppType>());
+            }
+        }
+
+        auto& dest_data_column = dest_column->elements_column();
+        auto& dest_offsets = dest_column->offsets_column()->get_data();
+
+        if (has_null) {
+            dest_data_column->append_nulls(1);
+        }
+
+        auto iter = hash_set->begin();
+        while (iter != hash_set->end()) {
+            dest_data_column->append_datum(*iter);
+            iter++;
```

Comment:
prefer ++iter to iter++

Reply:
It should be equivalent under O3

Reply:
Fixed

Reply:
I think we don't have this preference.

4. x = array[i++] is more efficient than x = array[++i]
5. x = *(p++) is more efficient than x = *(++p)

Reply:
This may not be the main factor affecting performance. It can be ignored for now. If the test improves a lot, I will change it.